manual/tracer.py.1.ronn
# tracer.py(1) -- use straight-ray tracing to estimate two-medium parameters

## SYNOPSIS

`tracer.py` <CONFIGURATION>


## DESCRIPTION

`tracer.py` uses the surface mesh produced by `wavesfc.py` to trace paths
through a two-medium model. The portions of each path in the interior and
exterior of the volume, combined with estimates of arrival times for waves
propagating along the paths, are used to estimate the interior and exterior
sound speeds in a least-squares sense.

The interior and exterior speeds may be treated as separable or linked in a
common optimization problem.

For detecting interior and exterior length fractions of each path, the volume
spanned by the surface mesh is decomposed with an octree to allow for fast
searching of intersections between paths and surface facets. Paths with an odd
number of surface intersections (i.e., those that enter the interior but never
leave) are discarded.


## CONFIGURATION

The YAML configuration file (or the rendered Mako template that produces valid
YAML) should define a dictionary with four sections. Each section is, in turn,
a dictionary that defines configuration options. Some sections are common to
multiple programs, allowing these programs to share a single configuration
file without duplicating common options.

The `measurement` section includes one relevant key:

- `c` (optional, default: _1.4823_):
  An initial estimate for the sound speed of the background medium surrounding
  the target.

The `tracer` section includes several options:

- `timefile`:
  A single path or glob, or a list of paths or globs, to keymat(7) files that
  map transmit-receive pairs to arrival times. Any transmit-receive pair for
  which the transmit and receive indices are not equal are discarded. If each
  transmit-receive pair maps to a list of more than one arrival time, one
  target positions per arrival time will be determined.

- `targidx` (optional, default: _0_):
  If `timefile` keymat(7) files specify more than one arrival time for each
  transmit-receive key, the value of `targidx` determines which time is used
  for sound-speed estimates.

- `elements`:
  A keymat(7) file that maps individual element indices (as single integers) to
  a location as an (x, y, z) array of floats.

- `mesh`:
  A Numpy *npz* file describing a surface mesh produced by `wavesfc.py`.

- `levels`:
  Decompose the volume spanned by the mesh into a `levels`-level octree.

- `vclip` (optional):
  If provided, a list [<MIN>, <MAX>] that specifies the minimum and maximum
  permissible average sound speed in the optimization problem. Arrival times
  that impute an average speed outside of the range will be discarded.

- `bimodal` (optional, default: _True_):
  If _True_, solve an optimization problem to recover exactly one exterior and
  one interior sound speed from all of the measurements. If _False_, determine
  a single exterior sound speed from paths that miss the target surface, and
  then compute individual estimates of interior sound speeds for each traced
  path that intersects the surface.

- `fixbg` (optional, default: _False_):
  If _True_, do not attempt to find an optimal background sound speed, instead
  using the fixed `measurement` value.

- `separable` (optional, default: _False_):
  If _True_, treat `bimodal` interior and exterior speeds as separable. The
  exterior sound speed will be estimated from paths that miss the target, and
  will then be held constant to recover a single, optimal interior sound speed
  from paths that intersect the surface.

  If _False_, paths that miss the target are used to provide a first guess of
  the exterior sound speed, but LSMR is used to find optimal interior and
  exterior sound speeds from paths that intersect the target.

  When `bimodal` is _False_, this option has no effect.
  
- `output`:
  The name of a file that will be written. If `pathsave` is _False_ and
  `bimodal` is _True_, `output` will be a text file that lists the interior
  and exterior sound speeds for the bimodal model.

  Otherwise, `output` will be a Numpy *npz* file with at least two constituent
  files: `exspd`, which is a one-element array that specifies the exterior
  sound speed; and `inspd`, which lists the interior speed(s) in a format that
  depends on whether `bimdal` is _True_.

  When `bimodal` is _True_, `inspd` is a one-element array the provides the
  single interior sound speed. Otherwise, `inspd` is a Numpy record array
  where each record provides a transmit index `tx`, a receive index `rx` and
  an interior sound speed `inspd` for that path.

- `pathsave` (optional, default: _False_):
  When _True_, `output` will always be a Numpy *npz* file. In addition to the
  standard constituent files, two additional constituent files will be added.
  Both constituent files describe individual paths as Numpy record arrays
  where each record provides a transmit index `tx`, a receive index `rx`, an
  exterior length `exlen`, an interior length `inlen`, and an associated
  arrival time `atime`. The first constituent file, `hits`, describes paths
  determined to have hit the target. The second constituent file, `misses`,
  describes those paths determined to have missed the target.

- `epsilon` (optional, default: _1.0e-3_):
  A tolerance such that any path with an interior-to-exterior length ratio
  less than `epsilon` will be considered to have missed the target.

- `mask_outliers` (optional, default: _False_):
  When _True_, arrival times from each input `timefile` will be removed if
  they fall more than 1.5 IQR above the third quartile or below the first
  quartile. Note that the IQR and quartiles are calculated on a per-file
  basis.


## COPYRIGHT

Copyright (c) 2014--2018 Andrew J. Hesford. All rights reserved.


## SEE ALSO

wavesfc.py(1) - For a description of the surface mapping process  
keymat(7) - For a description of the keymat file format


shell/tracer.py
#!/usr/bin/env python

# Copyright (c) 2016 Andrew J. Hesford. All rights reserved.
# Restrictions are listed in the LICENSE file distributed with this package.

import os, sys, math, numpy as np
from numpy.linalg import norm

from scipy.sparse.linalg import lsmr

import time

from mpi4py import MPI

from habis.habiconf import HabisConfigParser, HabisConfigError, matchfiles
from habis.formats import loadmatlist as ldmats, loadkeymat as ldkmat, savez_keymat
from habis.mpdfile import flocshare, getatimes

from pycwp import boxer, cutil


def maketimestr(start, end):
	'''
	Print the time (start - end) in seconds (with millisecond precision) if
	the time is greater than 1 second, or in milliseconds (with microsecond
	precision) if the time is less than 1 second. Add a suffix (s or ms).
	'''
	dtime = end - start
	if abs(dtime) < 1: return '%0.3f ms' % (1000. * dtime,)
	else: return '%0.3f s' % (dtime,)


def makedtypes(nidx=2, nvals=3, names=None):
	'''
	Create and return a Numpy dtype and corresponding MPI structure data
	type consists of nidx 64-bit long integers followed by nvals 64-bit
	floats. The Numpy dtype is the first return value.

	If names is provided, it must be a string of length (nidx + nvals), or
	a sequence of (nidx + nvals) strings, that will be used to name the
	fields of the Numpy dtype; if a single string is provided, the fields
	will be given names corresponding to the characters within.

	If names is not provided, the dtype fields will be anonymous (i.e.,
	they will be named 'f0', 'f1', etc.).
	'''
	# Create data types to store and share segment intersection results
	dt = np.dtype(','.join(['<i8']*nidx + ['<f8']*nvals))
	if names: dt.names = names
	offsets = [dt.fields[n][1] for n in dt.names]
	mtypes = [MPI.LONG]*nidx + [MPI.DOUBLE]*nvals
	mpt = MPI.Datatype.Create_struct([1]*len(mtypes), offsets, mtypes)
	mpt.Commit()
	return dt, mpt


def gatherleaves(otree, comm=None):
	'''
	Gather on MPI communicator comm (or MPI.COMM_WORLD if comm is None) the
	leaves from Octree objects on all ranks and merge into the Octree
	otree. The leaves of otree will be sent by this rank.

	Nothing is returned.
	'''
	if comm is None: comm = MPI.COMM_WORLD
	for leaves in comm.allgather(otree.getleaves()):
		otree.mergeleaves(leaves)


def gathersegments(segs, mptype, comm=None, root=0):
	'''
	Gather, on the rank such that comm.Get_rank() == root for the
	communicator comm (MPI.COMM_WORLD if comm is None), all records in the
	rank-1 structured Numpy array segs with a Numpy dtype compatible with
	the MPI data type mptype.
	'''
	if comm is None: comm = MPI.COMM_WORLD
	rank = comm.Get_rank()

	# Gather the count from each rank
	nsegs = len(segs)
	counts = comm.gather(nsegs)

	# Allocate space on the root for all records; other ranks don't care
	if rank == root:
		combined = np.empty((sum(counts),), dtype=segs.dtype)
		# Compute the offsets for data received from each rank
		displs = [0]
		for ct in counts[:-1]: displs.append(displs[-1] + ct)
	else:
		combined, displs = None, None

	comm.Gatherv([segs, nsegs, mptype], [combined, counts, displs, mptype])
	return combined


def usage(progname=None, fatal=True):
	if not progname: progname = os.path.basename(sys.argv[0])
	print('USAGE: %s <configuration>' % progname, file=sys.stderr)
	sys.exit(int(fatal))


def tracerEngine(config):
	'''
	Given lists of arrival-time maps, element positions, and lists of
	control points and triangles that define a target surface, all
	specified in the HabisConfigParser instance config, use a
	pycwp.boxer.Octree to identify points of intersections between the
	surface and each non-backscatter propagation path for which arrival
	times are defined. From these intersection points, deduce the fraction
	of each propagation path is spent inside the target, and determine an
	average interior sound speed from the arrival time and an assumed
	background sound speed.
	'''
	tsec = 'tracer'

	def _throw(msg, e, sec=tsec):
		raise HabisConfigError.fromException(msg + ' in [%s]' % (sec,), e)

	# Find all arrival-time maps visible to this node
	try: timefiles = matchfiles(config.getlist(tsec, 'timefile'), forcematch=False)
	except Exception as e: _throw('Configuration must specify timefile', e)

	try: targidx = config.get(tsec, 'targidx', mapper=int, default=0)
	except Exception as e: _throw('Invalid optional targidx', e)

	# Find and load all element coordinates
	try: elements = ldmats(matchfiles(config.getlist(tsec, 'elements')), nkeys=1)
	except Exception as e: _throw('Configuration must specify elements', e)

	# Find and load all propagation axes
	try: meshctr = np.loadtxt(config.get(tsec, 'meshctr'), ndmin=2)[targidx,:3]
	except Exception as e: _throw('Configuration must specify meshctr', e)

	# Load the node and triangle configuration
	try: meshfile = config.get(tsec, 'mesh')
	except Exception as e: _throw('Configuration must specify mesh', e)

	try: levels = config.get(tsec, 'levels', mapper=int)
	except Exception as e: _throw('Configuration must specify levels', e)

	# Read the background sound speed or use water at 68F by default
	try: vbg = config.get('measurement', 'c', mapper=float, default=1.4823)
	except Exception as e: _throw('Invalid optional c', e, 'measurement')

	try:
		# Pull a range of valid sound speeds for clipping
		vclip = config.getlist(tsec, 'vclip', mapper=float, default=None)
		if vclip:
			if len(vclip) != 2:
				raise ValueError('Range must specify two elements')
			if vclip[0] > vclip[1]:
				raise ValueError('Minimum value must not exceed maximum')
			vclip = tuple(vclip)

	except Exception as e:
		_throw('Invalid optional vclip', e)

	try: bimodal = config.get(tsec, 'bimodal', mapper=bool, default=True)
	except Exception as e: _throw('Invalid optional bimodal', e)

	try: fixbg = config.get(tsec, 'fixbg', mapper=bool, default=False)
	except Exception as e: _throw('Invalid optional fixbg', e)

	try: separable = config.get(tsec, 'separable', mapper=bool, default=False)
	except Exception as e: _throw('Invalid optional separable', e)

	try: pathsave = config.get(tsec, 'pathsave', mapper=bool, default=False)
	except Exception as e: _throw('Invalid optional pathsave', e)

	try: output = config.get(tsec, 'output')
	except Exception as e: _throw('Configuration must specify output', e)

	try: epsilon = config.get(tsec, 'epsilon', default=1e-3)
	except Exception as e: _throw('Invalid optional epsilon', e)

	try: mask_outliers = config.get(tsec, 'maskoutliers', default=False)
	except Exception as e: _throw('Invalid optional maskoutliers', e)

	WORLD = MPI.COMM_WORLD
	mpirank, mpisize = WORLD.rank, WORLD.size

	WORLD.Barrier()

	stime = time.time()

	# Convert the triangle node maps to triangle objects
	with np.load(meshfile) as mesh:
		nodes, triangles = mesh['nodes'], mesh['triangles']
	# Retain the unique labeling of nodes in all Triangle3D objects
	triangles = [ boxer.Triangle3D([nodes[c] for c in v], v) for v in triangles ]

	# Compute an overall bounding box and an Octree for the space
	rootbox = boxer.Box3D(*zip(*((min(j), max(j))
		for j in zip(*(nd for tr in triangles for nd in tr.nodes)))))
	otree = boxer.Octree(levels, rootbox)

	if not mpirank:
		print('Scatterer has %d triangles, %d nodes' % (len(triangles), len(nodes)))
		print('Limiting sound-speed values to range %s' % (vclip,))
		print('Creating Octree decomposition (%d levels)' % (otree.level,))

	# Classify triangles according to overlaps with boxes in tree
	def inbox(b, i): return triangles[i].overlaps(b)
	# Only build the tree for a local share of triangles
	otree.addleaves(range(mpirank, len(triangles), mpisize), inbox, True)

	etime = time.time()
	if not mpirank:
		print('Local tree build time:', maketimestr(stime, etime))

	WORLD.Barrier()

	if not mpirank: print('Combining distributed Octree')

	stime = time.time()

	# Merge trees and prune
	gatherleaves(otree)
	otree.prune()

	etime = time.time()
	if not mpirank:
		print('Tree aggregation time:', maketimestr(stime, etime))

	WORLD.Barrier()

	if not mpirank:
		print('Reading and gathering arrival times')
		if mask_outliers: print('Will exclude outlier arrival times')

	# Collect the arrival times from local shares of all local maps
	atimes = { }
	for tfile, (start, stride) in flocshare(timefiles, MPI.COMM_WORLD).items():
		atimes.update(getatimes(tfile, elements, targidx, True,
					vclip, mask_outliers, start, stride))

	if not mpirank: print('Approximate local share of paths is %d' % (len(atimes),))

	# Build the segment list for the local arrival times
	segs = { }
	for t, r in atimes:
		if t != r:
			segs[t,r] = boxer.Segment3D(elements[t], elements[r])
			continue
		# For backscatter, make segment length encompass volume
		epos = elements[t]
		segs[t,t] = boxer.Segment3D(epos, meshctr)

	WORLD.Barrier()

	stime = time.time()

	# Accumulate the total results for every segment
	dt, mpt = makedtypes(2, 3, ('tx', 'rx', 'exlen', 'inlen', 'atime'))
	results = np.empty((len(segs),), dtype=dt)
	nres = 0
	# Track the number of odd-intersection paths that are skipped
	nodds = 0

	# Find the portion of each path in the interior and exterior of the volume
	for k, seg in segs.items():
		# Skip very small segments
		if cutil.almosteq(seg.length, 0.0, epsilon): continue

		# Predicates match segment-box and segment-triangle intersections
		def bsect(b): return b.intersection(seg)
		def tsect(i): return triangles[i].intersection(seg)

		# Find intersections between segment and surface; cache
		# segment-triangle intersections to avoid redundant tests
		isects = otree.search(bsect, tsect, { })

		if k[0] == k[1]:
			# Backscatter: nonsense if there is no surface intersection
			if not len(isects): continue
			# Exterior path is to first hit and back, no interior
			exl = 2.0 * min(v[0] for v in isects.values())
			inl = 0.0
		else:
			# Through transmissions: add endpoints to define all regions
			ilens = sorted([0., 1.0] +
					[v[0] for v in isects.values()])

			# Odd count: segment starts or ends in interior
			if len(ilens) % 2:
				nodds += 1
				continue

			# Interior regions start at odd indices, exterior on evens
			inl = sum(v[1] - v[0] for v in zip(ilens[1::2], ilens[2::2]))
			exl = sum(v[1] - v[0] for v in zip(ilens[0::2], ilens[1::2]))

			# Check total length against segment length (ratio)
			if not cutil.almosteq(inl + exl, 1.0, epsilon):
				print(('WARNING: (t,r) segment %s inferred to '
					'actual ratio %0.5g to 1' % (k, inl + exl)))

		# In fixed-exterior mode, ignore all-exterior segments
		if fixbg and abs(inl) <= epsilon * abs(exl): continue

		# Convert length from fractions of total to actual lengths
		exl *= seg.length
		inl *= seg.length

		# Capture the arrival time for this segment
		atime = atimes[k]

		# Check speeds for sanity, if limits have been specified
		if vclip:
			if abs(inl) <= epsilon * abs(exl):
				# Propagation is all "exterior"
				vx = exl / atime
				if not vclip[0] <= vx <= vclip[1]: continue
			else:
				# Assume exterior speed is nominal value
				tex = exl / vbg
				if cutil.almosteq(atime, tex, epsilon): continue
				vi = inl / (atime - tex)
				if not vclip[0] <= vi <= vclip[1]: continue

		results[nres] = k + (exl, inl, atime)
		nres += 1

	etime = time.time()

	# Accumulate the list of skipped segments on the root
	nodds = WORLD.reduce(nodds)
	if not mpirank:
		if nodds:
			print('Skipped %d paths with an odd intersections count' % (nodds,))
		print('Local segment tracing time:', maketimestr(stime, etime))

	WORLD.Barrier()

	# Accumulate all results on the root
	results = gathersegments(results[:nres], mpt)
	# No need for the MPI type anymore
	mpt.Free()

	# Only the root has any work left
	if mpirank: return

	print('Finished tracing; attempting to determine speeds')

	stime = time.time()

	# Split tracing results into "hits" and "misses"
	hits = np.abs(results['inlen']) > epsilon * np.abs(results['exlen'])
	misses = results[np.logical_not(hits)]
	hits = results[hits]

	if not fixbg and len(misses):
		# Update the exterior speed, if desired and possible
		t = misses['atime']
		x = misses['exlen']
		vbg = (t @ x) / (t @ t)

	print('Exterior speed from misses: %0.5g (%d paths)' % (vbg, len(misses)))

	# Offset arrival times by contribution from exterior speed
	rhs = hits['atime'] - hits['exlen'] / vbg

	if bimodal:
		if separable or fixbg:
			# If separable or fixed background,
			# bimodal system is a single-column matrix
			x = ((rhs @ hits['inlen']) / (rhs @ rhs), )
		else:
			# If not separable, bimodal system is a two-column matrix
			A = np.array([hits['exlen'], hits['inlen']]).T
			x, istop, itn, *tols = lsmr(A, rhs)
			if istop > 2:
				print('LSMR terminated with stop code %d at iteration %d' % (istop, itn))
			# First component is difference in exterior slowness
			vbg = vbg / (vbg * x[0] + 1)
			# Second component is whole interior slowness
			x = (1 / x[1],)
			print('Exterior speed from optimization: %0.5g (%d paths)' % (vbg, len(hits)))
	else:
		# Multimodal matrix is diagonal
		x = tuple((hv / rv) if abs(rv) > epsilon else 0.
				for hv, rv in zip(hits['inlen'], rhs))

	# If Npz output will be used, always save the exterior speed
	szargs = { 'exspd': (vbg,) }

	if bimodal:
		print('Interior speed: %0.5g (%d paths)' % (x[0], len(hits)))
		if not pathsave:
			# Write a simple text file if paths aren't saved
			with open(output, 'wb') as f:
				f.write(b'# Exterior sound speed\n')
				f.write(b'%0.18g\n' % (vbg,))
				f.write(b'# Interior sound speed\n')
				f.write(b'%0.18g\n' % (x[0],))
			return
		# Store the interior speed is a single value (as a 1-element array)
		szargs['inspd'] = x
	else:
		# Print some useful stats on the recovered values
		stats = np.mean(x), np.median(x), np.std(x), len(x)
		print(('Interior speed: mean %0.5g, '
				'median %0.5g, std %0.5g (%d paths)' % stats))

		# Store the tx, rx, and speed values in multipath mode
		ist = [('tx', '<i8'), ('rx', '<i8'), ('inspd', '<f8')]
		ispds = np.empty(hits.shape, dtype=ist)
		ispds['tx'] = hits['tx']
		ispds['rx'] = hits['rx']
		ispds['inspd'] = x
		szargs['inspd'] = ispds

	etime = time.time()
	if not mpirank:
		print('Speed determination time:', maketimestr(stime, etime))


	if pathsave:
		# Store the hits and misses as paths only, if they exist
		if len(hits): szargs['hits'] = hits
		if len(misses): szargs['misses'] = misses

	# Save the output
	print('Saving results')
	np.savez(output, **szargs)


if __name__ == '__main__':
	if len(sys.argv) != 2:
		usage()

	try:
		config = HabisConfigParser(sys.argv[1])
	except Exception as e:
		err = 'Unable to load configuration file %s' % sys.argv[1]
		raise HabisConfigError.fromException(err, e)

	try:
		tracerEngine(config)
	except Exception as e:
		rank = MPI.COMM_WORLD.rank
		print('MPI rank %d: caught exception %s' % (rank, e))
		sys.stdout.flush()
		MPI.COMM_WORLD.Abort(1)
